### PR TITLE
[feature] Support max_query_dependency_size

### PR DESCRIPTION
**Pull request by donaldong**
### Summary
This adds a feature to respect `max_query_dependency_size`.

Users might have big queries with a super big dependency list -- if we perform cross-product on those dependencies, it might create millions of entries. Thus, this adds a way to "plan" first, before actually extracting the bind_params.

This also adds a tracer span, so we can see the time spent in extracting dependies.

### Test Plan
- Added test cases

### CODE REVIEW

File: lib/redis_memo/memoize_query/cached_select.rb [hunk 1]
```diff
@@ -318,9 +318,6 @@ def self.extract_bind_params_recurse(node)
 
           return unless bind_params
         end
-
-        # Reject any unbound select queries
-        return if binding_relation && bind_params.params[binding_relation].empty?
```
Comment (donaldong) [inline lib/redis_memo/memoize_query/cached_select.rb:323]: Moved to `BindParams#should_cache?`

File: lib/redis_memo/memoize_query/cached_select.rb [hunk 2]
```diff
@@ -167,20 +167,20 @@ def map_substitutes(values)
...
-    RedisMemo::ThreadLocalVar.arel_bind_params = bind_params
-    true
+    RedisMemo::Tracer.trace('redis_memo.memoize_query.extract_bind_params', nil) do
+      ast = RedisMemo::ThreadLocalVar.arel&.ast
+      return false unless ast.is_a?(Arel::Nodes::SelectStatement)
+      return false unless ast.to_sql == sql
+
+      RedisMemo::ThreadLocalVar.substitues ||= {}
+      # Iterate through the Arel AST in a Depth First Search
+      bind_params = extract_bind_params_recurse(ast)
```
Comment (katyho) [inline lib/redis_memo/memoize_query/cached_select.rb:177]: How about renaming `extract_bind_params_recurse`? Maybe to  `construct_bind_params_recurse` or something similar, but since we don't actually extract the params until `extract!` is called

**Review by katyho — COMMENTED** [review]
I did a first pass (not including unit tests) and found some of the logic a bit hard to follow. Added some comments for clarification

File: lib/redis_memo/memoize_query/cached_select.rb (hunk 2 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select.rb:177]: Good point!

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 3]
```diff
@@ -2,47 +2,107 @@
...
+
+        attrs_set.each do |attrs|
+          return false unless RedisMemo::MemoizeQuery
+            .memoized_columns(model)
+            .include?(attrs.keys.sort)
         end
       end
 
-      other
+      !plan.model_attrs.empty?
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:38]: Sure!

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 4]
```diff
@@ -2,47 +2,107 @@
...
+
+    protected
+
+    attr_accessor :left
+    attr_accessor :right
+    attr_accessor :opt
+    attr_accessor :plan
+
+    def union_attrs_set(left, right)
+      left.merge(right) do |_, attrs_set, other_attrs_set|
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:85]: This is only relevant when we're doing in-place merging.  Now we're creating brand new nodes, copying over all the keys and values, so the order does not matter anymore.

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 5]
```diff
@@ -2,47 +2,107 @@
...
+    def plan_union
+      plan.dependency_size = left.plan.dependency_size + right.plan.dependency_size
+      plan.model_attrs = union_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
+
+    def union!
+      @params = union_attrs_set(left.params, right.params)
+    end
+
+    def product_attrs_set(left, right)
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:102]: The content we're multiplying is a bit different.

In the planning phase, the `attrs_set`s always have a `nil` value as the hash content.

For example:
```ruby
{a:nil} x {b: nil} => {a: nil, b: nil}
{a:nil, b:nil} x {a: nil: b: nil} => {a: nil, b: nil}
```

When in the extraction phase, that's where the # of dependency can actually blow up:
```ruby
{a: [1,2,3]} x {b: [1,2,3]} => [{a: 1, b: 1}, ....]
{a:[1,2], b:[1,2]} x {a: [1,2,3]: b: [1,2,3]} => [...]
```

Technically `plan.model_attrs` can also just be a `Set` instead of `Hash`. But the performance is about the same, and we do have to write another code path just to handling  adding/multiplying sets

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 6]
```diff
@@ -55,73 +115,68 @@ def product(other)
...
+      plan.dependency_size = left.plan.dependency_size * right.plan.dependency_size
+      plan.model_attrs = product_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
     end
 
-    def uniq!
-      params.each do |_, attrs_set|
-        attrs_set.uniq!
-      end
+    def product!
+      @params = product_attrs_set(left.params, right.params)
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:156]: `@params` contains the ready-to-use result after calling `extract!`

`plan.model_attrs` is a simplified `@params`, that values of each attr is ignored.

See also https://github.com/chanzuckerberg/redis-memo/pull/94/files#r650258999


>  It's confusing to me that we're resetting it here.

We're only setting it for the nodes that have an `operation`. The leaf nodes do not have operations, and only leaf nodes contain the original params from the Arel query.

We are setting the `@params` here to construct the final `@params` recursively. When we are looking at the parent nodes, the params from the left and right child are ready to use, we can just multiply them together.

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 7]
```diff
@@ -2,47 +2,107 @@
...
-      #   }
-      #
-      @params ||= Hash.new do |models, model|
-        models[model] = []
-      end
+    def initialize(left = nil, right = nil, opt = nil)
+      @left = left
+      @right = right
+      @opt = opt
+      @plan = nil
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:9]: can we comment and explain here that the plan is used to generate a plan and determine whether to cache the query?

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 8]
```diff
@@ -2,47 +2,107 @@
...
+        next other_attrs_set if attrs_set.empty?
+
+        attrs_set + other_attrs_set
+      end
+    end
+
+    def plan_union
+      plan.dependency_size = left.plan.dependency_size + right.plan.dependency_size
+      plan.model_attrs = union_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:96]: Can we make `Plan` own the `union` method by passing in two `plan` obejcts?

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 9]
```diff
@@ -55,73 +115,68 @@ def product(other)
...
-        merged_attrs_set.compact!
-        other.params[model] = merged_attrs_set
+        merged_attrs_set
       end
+    end
 
-      other
+    def plan_product
+      plan.dependency_size = left.plan.dependency_size * right.plan.dependency_size
+      plan.model_attrs = product_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:152]: also here, i think `plan` can own the `product` method.

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 10]
```diff
@@ -55,73 +115,68 @@ def product(other)
...
-            .memoized_columns(model)
-            .include?(attrs.keys.sort)
+        # An aggregated bind_params node can only obtain params by combining
+        # its children nodes
+        return if !bind_params.__send__(:opt).nil?
+
+        bind_params.params.each do |model, attrs_set|
+          @dependency_size += attrs_set.size
+          attrs_set.each do |attrs|
+            @model_attrs[model] << attrs.keys.map { |k| [k, nil] }.to_h
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:176]: can we explain why we passed in `nil` here? also related to https://github.com/chanzuckerberg/redis-memo/pull/94/files#r650273177, I think using a set seems to make more sense here, so if the advantage of using a Hash here is to avoid another code path, I think we should add some explanations here

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 11]
```diff
@@ -55,73 +115,68 @@ def product(other)
...
               #
               #   (a = 1 or b = 1) and (a = 2 or b = 2)
               #
-              #  Keep:     a = 1 and b = 2, a = 2 and b = 1
-              #  Discard:  a = 1 and a = 2, b = 1 and b = 2
+              #  keep:     a = 1 and b = 2, a = 2 and b = 1
+              #  discard:  a = 1 and a = 2, b = 1 and b = 2
               if merged_attrs.include?(name) && merged_attrs[name] != val
-                merged_attrs_set[k] = nil
+                should_add = false
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:136]:
```suggestion
                should_add_attr = false
```

to be more concrete

**Review by Dingying0410 — COMMENTED** [review]
mostly makes sense to me, and I think more comments and explanations on what the `plan` does will be nice!

do we also need to add the doc for this particular feature?

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb (hunk 8 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:96]: A `Plan` is a simple data hold object. It does not really makes sense for it to own `union`  since this method is used outside of the planning phase as well (see `union!`)

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb (hunk 9 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:152]: similar to the above comment, product is used outside of the planning phase too

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 12]
```diff
@@ -55,73 +115,68 @@ def product(other)
...
-      params.each do |_, attrs_set|
-        attrs_set.uniq!
-      end
+    def product!
+      @params = product_attrs_set(left.params, right.params)
     end
 
-    def memoizable?
-      return false if params.empty?
+    class Plan
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:159]: Yeah I will add some explanation for what the is planning phase is doing in general!

**Review by katyho — CHANGES_REQUESTED** [review]
I feel like there might be a bug somewhere in traversing the actual tree and computing the plan size.

I checked out your branch, and tried to do 
`Student.where(id: (1..5000).to_a).where(id: (1..5000).to_a)`, but this does end up getting cached on RedisMemo even with the max dependency size set to something relatively small. The plan dependency ends up being 0 because at some point we're multiplying 0*5000

**Comment by donaldong** [comment]
> The plan dependency ends up being 0 because at some point we're multiplying 0*5000

uhhhh, i see. Yeah the logic for calculating denepdency_size in this PR is just wrong 😅  will add a test case covering this senerio

File: lib/redis_memo/memoize_query/cached_select.rb [hunk 13]
```diff
@@ -162,25 +162,25 @@ def map_substitutes(values)
...
-    # Iterate through the Arel AST in a Depth First Search
-    bind_params = extract_bind_params_recurse(ast)
-    return false unless bind_params
-
-    bind_params.uniq!
-    return false unless bind_params.memoizable?
-
-    RedisMemo::ThreadLocalVar.arel_bind_params = bind_params
-    true
+    RedisMemo::Tracer.trace('redis_memo.memoize_query.extract_bind_params', nil) do
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select.rb:170]: Ah, that's a good idea!

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 14]
```diff
@@ -2,47 +2,217 @@
...
+
+    def plan_union
+      plan.dependency_size_estimation = left.plan.dependency_size_estimation + right.plan.dependency_size_estimation
+      plan.model_attrs = union_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
+
+    def plan_product
+      plan.dependency_size_estimation = left.plan.dependency_size_estimation * right.plan.dependency_size_estimation
+      plan.model_attrs = product_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:193]: from line 185 until line 201, I would expect something that some obvious distinction for the planning phase and extraction phase thing. I think we can at least we can use comments or method names to specify.

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 15]
```diff
@@ -2,47 +2,217 @@
...
+
+    def union!
+      @params = union_attrs_set(left.params, right.params)
+    end
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:197]: if this union is now used in the extraction, can we name it as `extrsction_union`?

File: spec/memoize_query/cached_select/bind_params_spec.rb [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    left = described_class.new
+    left.params[1] << { a: 1 }
+
+    right = described_class.new
+    right.params[1] << { b: 2 }
+
+    bp = left.product(right)
+    bp.extract!
+    expect(bp.params[1].to_a).to eq([{ a: 1, b: 2 }])
+  end
```
Comment (Dingying0410) [inline spec/memoize_query/cached_select/bind_params_spec.rb:28]: can we add a test case here where either left or right is empty?

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 17]
```diff
@@ -2,47 +2,217 @@
...
+        return if !bind_params.__send__(:operator).nil?
+
+        bind_params.params.each do |model, attrs_set|
+          @dependency_size_estimation[model] += attrs_set.size
+          attrs_set.each do |attrs|
+            # [k, nil]: Ignore the attr value and keep the name only
+            @model_attrs[model] << attrs.keys.map { |k| [k, nil] }.to_h
+          end
+        end
+      end
```
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:171]: separate question from the code: do we have an estimate here how much time will take to compute the plan dependency size? how does it scale when the attr set is large.

**Review by Dingying0410 — COMMENTED** [review]
Thanks! The code and logic itself seems reasonable to me so far, the added comments are more clear, left some comments which i think will make the logic more clear

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 18]
```diff
@@ -2,47 +2,217 @@
...
+    end
+
+    def should_cache?
+      plan!
+
+      if plan.model_attrs.empty? || plan.dependency_size_estimation.to_i > RedisMemo::DefaultOptions.max_query_dependency_size
+        return false
+      end
+
+      plan.model_attrs.each do |model, attrs_set|
```
Comment (katyho) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:30]: Maybe a micro-optimization, but I wonder if we can stop traversing through the AST if we hit a column that isn't memoized and exit early, rather than first having to `plan!` (though I guess the costly logic is in the `extract!`)

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 19]
```diff
@@ -2,47 +2,217 @@
...
+    #   {a:nil} x {b: nil} => {a: nil, b: nil}
+    #   {a:nil, b:nil} x {a: nil: b: nil} => {a: nil, b: nil}
+    #
+    # and in the extraction phase, that's where the # of dependency can
+    # actually grow significantly:
+    #
+    #   {a: [1,2,3]} x {b: [1,2,3]} => [{a: 1, b: 1}, ....]
+    #   {a:[1,2], b:[1,2]} x {a: [1,2,3]: b: [1,2,3]} => [{a: 1, b: 1}, ...]
+    #
+    class Plan
```
Comment (katyho) [inline lib/redis_memo/memoize_query/cached_select/bind_params.rb:102]: nit: Could you move this class to the bottom of the file? Unless there's a good reason to put it here, I found it a bit hard to read since it breaks up the `plan!` and `extract!` methods.

Also maybe put all the attr_accessors of the BindParam class together

File: spec/memoize_query/cached_select/bind_params_spec.rb [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    bp = left.union(right)
+    bp.extract!
+    expect(bp.params[1].to_a).to eq([{ a: 1 }, { b: 2 }])
+
+    left.params[1] << { b: 3 }
+    bp.extract!
+    expect(bp.params[1].to_a).to eq([{ a: 1 }, { b: 3 }, { b: 2 }])
+  end
+
+  it 'products' do
```
Comment (katyho) [inline spec/memoize_query/cached_select/bind_params_spec.rb:18]: I think a more complicated test case with a query that contains both a product and union would be beneficial

File: spec/memoize_query/cached_select/bind_params_spec.rb [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+
+    right = described_class.new
+    right.params[1] << { b: 2 }
+
+    bp = left.product(right)
+    bp.extract!
+    expect(bp.params[1].to_a).to eq([{ a: 1, b: 2 }])
+  end
+
+  it 'excludes conflict query conditions' do
```
Comment (katyho) [inline spec/memoize_query/cached_select/bind_params_spec.rb:30]: I might have asked this before 😅  what happens to a SQL query if there's a conflicting query condition and it hits the real database? What gets returned?

**Review by katyho — COMMENTED** [review]
Overall LGTM! Left a few mostly nit comments

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb [hunk 22]
```diff
@@ -2,47 +2,217 @@
...
+
+    def union!
+      @params = union_attrs_set(left.params, right.params)
+    end
+
+    def product!
+      @params = product_attrs_set(left.params, right.params)
+    end
+
+    def union_attrs_set(left, right)
```
[the remark this answers is not in the document]
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:203]: actually, union is the action here, and attrs_set is the subject.

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb (hunk 14 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:193]: Hmm as in one is + and the other is *? It feels pretty readable already though

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb (hunk 17 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:171]: Well, the attrs_set can never be that big. that's bounded by the length of different columns in the SQL query.  Calculating plan dependency size is an O(n) operation where n is the length of the SQL query

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb (hunk 15 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:197]: I think the current name makes a lot more sense to me. As in we are planning for an action, before actually perform an action

File: lib/redis_memo/memoize_query/cached_select/bind_params.rb (hunk 18 above)
Comment (donaldong) [reply lib/redis_memo/memoize_query/cached_select/bind_params.rb:30]: I thought about that -- does not seem practical though, because we don't know the exact set of attrs (e..g the id,name attr set) until the full AST has been passed.

E.g. A set of attrs might become cachable later, even if it is not currently at the current AST node

File: spec/memoize_query/cached_select/bind_params_spec.rb (hunk 21 above)
Comment (donaldong) [reply spec/memoize_query/cached_select/bind_params_spec.rb:30]: Oh, it just won't get return anything from the database server -- when performing the query, some query analyzer might even just take out the contradicting part automatically before executing the query

**Review by katyho — APPROVED** [review]

**Review by Dingying0410 — APPROVED** [review]
